src/envoy/transcoding/config.cc

```diff
@@ -0,0 +1,167 @@
+
+  ZeroCopyInputStream* ResponseOutput() { return response_stream_.get(); }
+  Status ResponseStatus() { return response_translator_->Status(); }
+
+ private:
+  std::unique_ptr<JsonRequestTranslator> request_translator_;
+  std::unique_ptr<ResponseToJsonTranslator> response_translator_;
+  std::unique_ptr<TranscoderInputStream> request_stream_;
+  std::unique_ptr<TranscoderInputStream> response_stream_;
+};
+}
+
+Config::Config(const Json::Object& config, Server::Instance& server) {
+  std::string proto_descriptor_file = config.getString("proto_descriptor");
+  std::fstream input(proto_descriptor_file, std::ios::in | std::ios::binary);
+  FileDescriptorSet descriptor_set;
+  if (!descriptor_set.ParseFromIstream(&input)) {
+    throw EnvoyException("Unable to parse proto descriptor");
+  }
+  log().debug("transcoding filter loaded");
+  for (const auto& file : descriptor_set.file()) {
+    if (descriptor_pool_.BuildFile(file) == nullptr) {
+      throw EnvoyException("Unable to parse proto descriptor");
+    }
+  }
+  resolver_.reset(google::protobuf::util::NewTypeResolverForDescriptorPool(
+      kTypeUrlPrefix, &descriptor_pool_));
+  info_.reset(google::protobuf::util::converter::TypeInfo::NewTypeInfo(
+      resolver_.get()));
+}
+
+Status Config::CreateTranscoder(const Http::HeaderMap& headers,
+                                ZeroCopyInputStream* request_input,
+                                TranscoderInputStream* response_input,
+                                std::unique_ptr<Transcoder>* transcoder) {
+  std::string path = headers.Path()->value().c_str();
+
+  auto method = ResolveMethod(headers.Method()->value().c_str(),
+                              headers.Path()->value().c_str());
+  if (!method) {
+    return Status(Code::NOT_FOUND,
+                  "Could not resolve " + path + " to a method");
+  }
+
+  RequestInfo request_info;
+  auto status = MethodToRequestInfo(method, &request_info);
+  if (!status.ok()) {
+    return status;
+  }
+
+  std::unique_ptr<JsonRequestTranslator> request_translator{
+      new JsonRequestTranslator(resolver_.get(), request_input, request_info,
+                                method->client_streaming(), true)};
+
+  auto response_type_url =
+      kTypeUrlPrefix + "/" + method->output_type()->full_name();
+  std::unique_ptr<ResponseToJsonTranslator> response_translator{
+      new ResponseToJsonTranslator(resolver_.get(), response_type_url,
+                                   method->server_streaming(), response_input)};
+
+  transcoder->reset(new TranscoderImpl(std::move(request_translator),
+                                       std::move(response_translator)));
+  return Status::OK;
+}
+
+Status Config::MethodToRequestInfo(
+    const google::protobuf::MethodDescriptor* method,
+    google::api_manager::transcoding::RequestInfo* info) {
+  // TODO: support variable bindings
+
+  auto request_type_url =
+      kTypeUrlPrefix + "/" + method->input_type()->full_name();
```

Comment:
so response_type_url is the same as request_type_url?

If so, can the code be shared?